src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/DirectiveAttributeCompletionItemProvider.cs

```diff
@@ -95,315 +93,324 @@ static bool InSnippetContext(
         }
     }
 
     private static void AddParameterNameCompletions(
-        TagHelperDescriptor descriptor,
-        BoundAttributeDescriptor attributeDescriptor,
-        ImmutableArray<(BoundAttributeDescriptor, string)> indexerDescriptors,
-        DirectiveAttributeCompletionContext context,
-        Dictionary<string, (ImmutableArray<BoundAttributeDescriptionInfo>, ImmutableArray<RazorCommitCharacter>, RazorCompletionItemKind)> attributeCompletions)
+        BoundAttributeDescriptor attribute,
+        ReadOnlySpan<BoundAttributeDescriptor> indexerAttributes,
+        DirectiveAttributeCompletionContext completionContext,
+        Dictionary<string, AttributeCompletionDetails> attributeCompletions)
     {
-        if (context.InAttributeName && !attributeDescriptor.IndexerNamePrefix.IsNullOrEmpty())
+        if (completionContext.InAttributeName && !attribute.IndexerNamePrefix.IsNullOrEmpty())
         {
             // Don't add parameters on indexers in attribute name contexts
             return;
         }
-        else if (context.InParameterName && !TagHelperMatchingConventions.CanSatisfyBoundAttribute(context.SelectedAttributeName, attributeDescriptor))
+
+        if (completionContext.InParameterName && !completionContext.CanSatisfyAttribute(attribute))
         {
             // Don't add parameters when the selected attribute name can't satisfy the given attribute descriptor in parameter name contexts
             return;
         }
 
         // Add indexer parameter completions first so they display first in completion descriptions.
-        foreach (var (indexerDescriptor, parentTagHelperTypeName) in indexerDescriptors)
+        foreach (var indexerAttribute in indexerAttributes)
         {
-            if (!attributeDescriptor.Name.StartsWith(indexerDescriptor.IndexerNamePrefix!))
+            var indexerNamePrefix = indexerAttribute.IndexerNamePrefix.AssumeNotNull();
+
+            if (!attribute.Name.StartsWith(indexerNamePrefix))
             {
                 continue;
             }
 
-            AddCompletionsForParameters(indexerDescriptor.Parameters, descriptor, attributeDescriptor, parentTagHelperTypeName, context, attributeCompletions);
+            AddCompletionsForParameters(attribute, indexerAttribute.Parameters, completionContext, attributeCompletions);
         }
 
         // Then add regular parameter completions
-        AddCompletionsForParameters(attributeDescriptor.Parameters, descriptor, attributeDescriptor, descriptor.TypeName, context, attributeCompletions);
+        AddCompletionsForParameters(attribute, attribute.Parameters, completionContext, attributeCompletions);
 
         return;
 
         static void AddCompletionsForParameters(
-            ImmutableArray<BoundAttributeParameterDescriptor> parameterDescriptors,
-            TagHelperDescriptor descriptor,
-            BoundAttributeDescriptor attributeDescriptor,
-            string parentTagHelperTypeName,
```

Comment:
>string parentTagHelperTypeName, [](http://example.com/codeflow?start=12&length=31)

I'm not exactly clear on how this isn't needed any more, but I'll trust that you verified it's functionally the same